added TPCircularBufferWithOffset

mlostekk opened the pull request:
added reading offset

mlostekk committed:
added TPCircularBufferWithOffset

mlostekk committed:
updated function documentation

TGabor commented on TPCircularBuffer.h:
```diff
@@ -143,6 +143,25 @@ static __inline__ __attribute__((always_inline)) void* TPCircularBufferTail(TPCi
+/*!
+ * Access end of buffer with a given offset
+ *
+ *  This gives you a pointer to the end of the buffer plus a given offset, ready
```
*  This gives you a pointer to the end of the buffer plus a given offset, ready
```suggestion
 *  This gives you a pointer to the tail of the buffer plus a given offset, ready
```

TGabor commented on TPCircularBuffer.h:
```diff
@@ -143,6 +143,25 @@ static __inline__ __attribute__((always_inline)) void* TPCircularBufferTail(TPCi
+ * @return Pointer to the first bytes ready for reading, or NULL if buffer is empty
+ */
+static __inline__ __attribute__((always_inline)) void* TPCircularBufferTailWithOffset(TPCircularBuffer *buffer, uint32_t offsetBytes, uint32_t* availableBytes) {
+    uint32_t bytesLeft = buffer->fillCount - offsetBytes;
```
`bytesLeft` is `uint32_t`, thus if `offsetBytes > fillcount`, we'll get an underflow

TGabor commented on TPCircularBuffer.h:
```diff
@@ -143,6 +143,25 @@ static __inline__ __attribute__((always_inline)) void* TPCircularBufferTail(TPCi
+static __inline__ __attribute__((always_inline)) void* TPCircularBufferTailWithOffset(TPCircularBuffer *buffer, uint32_t offsetBytes, uint32_t* availableBytes) {
+    uint32_t bytesLeft = buffer->fillCount - offsetBytes;
+    if (bytesLeft <= 0) return NULL;
+    *availableBytes = bytesLeft;
```
This should be set to 0 even if we don't return anything

TGabor committed:
Update availableBytes  and correct possible underflow in TPCircularBufferTailWithOffset

TGabor reviewed (commented):
I've rewritten parts of the function  in 
8ded3fe and 56a26d3 that address 
 * possiblity of underflow 
 * unset availableByes.
if `offsetBytes > fillcount`

@mlostekk could you have another look?

TGabor committed:
Fix availableBytes computation TPCircularBufferTailWithOffset

Only return something if the given offsetBytes is smaller than the buffer's fillCount.

mlostekk committed:
Update TPCircularBuffer.h

Co-Authored-By: Thomas Prätzlich <TGabor@users.noreply.github.com>
